php/web/index.php

```diff
@@ -15,6 +15,37 @@ function capture_output(callable $callable) {
     return $contents;
 }
 
+
+/**
+ * Makes a callable routine exclusively locked, aka "synchronized".
+ *
+ * @param string $lockname
+ *   The name of the lock. Will be used as a filename.
+ * @param callable $callable
+ *   The routine to make synchronous.
+ * @throws RuntimeException
+ *   If the lock failed for some reason.
+ * @return mixed
+ */
+function lock_exclusive(string $lockname, callable $callable) {
+    // Hash the filename to ensure it cannot contain weird path-manipulating characters.
+    $lockFile = __DIR__ . '/../locks/' . md5($lockname);
+    $fp = fopen($lockFile, 'w');
+    if ($fp === false) {
+        throw new RuntimeException(sprintf("Failed to open lock file: %s", escapeshellarg($lockname)));
+    }
+    if (!flock($fp, LOCK_EX)) {
+        throw new RuntimeException(sprintf("Failed to acquire lock: %s", escapeshellarg($lockname)));
+    }
+    try {
+        return $callable();
+    }
+    finally {
+        flock($fp, LOCK_UN);
+        fclose($fp);
```

Comment:
I believe just `fclose` is enough, isn't?
> On versions of PHP before 5.3.2, the lock is released also by fclose() (which is also called automatically when script finished).
https://www.php.net/manual/en/function.flock.php

Reply:
Before 5.3.2 that's the case.  We're running on 7.2 here, where by implication it would not be the case.  So we have to unlock it and then close.